### ISSUE TITLE
Inappropriate ioctl for device

### ISSUE THREAD

**Issue by jborden** [lajg-dev/Obsidian-Plugin-GPG-Inline-Encrypt#3]
Thank you for writing this plugin, I really like the idea of offloading encryption to the tried and true GPG utils. However, I am having an issue when trying to actually encrypt a file. 

I am running on: 
macOS 14.4.1 (23E224)
```
> gpg --version
gpg (GnuPG) 2.4.5
libgcrypt 1.10.3
```

Here is my config:
![GPG_Encrypt_Settings](https://github.com/lajg-dev/Obsidian-Plugin-GPG-Inline-Encrypt/assets/1325550/77230628-684b-4bfc-921c-5a944e25c8de)

Here is the actual message:
![Screenshot 2024-04-28 at 11 46 48 PM](https://github.com/lajg-dev/Obsidian-Plugin-GPG-Inline-Encrypt/assets/1325550/f067a63b-08e0-4be1-9e7f-63464ba0e873)

Here is the behavior:
![GPG_encryt_bug](https://github.com/lajg-dev/Obsidian-Plugin-GPG-Inline-Encrypt/assets/1325550/650f3c9c-d564-4302-b923-f2da1e913f8f)

What I've found so far:
(may or may not be relevant)
https://stackoverflow.com/questions/72445825/gpg-error-inappropriate-ioctl-for-device

**Comment by lajg-dev** [comment]
Hi @jborden, thanks for using this plugin.

I've been looking into your case, and it doesn't seem like these environment variable settings need to be set in the plugin, but rather they need to be set in your Mac's environment file (For example ~/.bashrc assuming bash)

Please see https://unix.stackexchange.com/a/257065

Please let me know if you managed to set it in your environment file and if this solved the problem for you

**Comment by jborden** [comment]
Hi thanks for looking into this. I did try adding the entries from the SO post. I tried adding entries with
```
GPG_TTY=$(tty)
export GPG_TTY
``` 
in ~/.profiles, ~/.bashrc, ~/.bash_profile, but it didn't solve the issue. I'm not sure how calls to command line tools works in Obsidian plugins and what environment they load. Perhaps looking at: https://stackoverflow.com/questions/72445825/gpg-error-inappropriate-ioctl-for-device/72788147#72788147, we could have an option for setting environment vars next to the path of the GPG Executable config option? The simplest would be if it could accept `export GPG_TTY=$(tty) ; /opt/homebrew/bin/gpg` 

Just FYI, I can run `gpg` commands at my terminal. 

Thanks!

**Comment by lajg-dev** [comment]
I will work on adding a section with custom commands to run before the gpg command, I will let you know when this change is ready, so you can try it out and see if it fixes your problem.

**Comment by lajg-dev** [comment]
Please download last version [1.1.2](https://github.com/lajg-dev/Obsidian-Plugin-GPG-Inline-Encrypt/releases/tag/1.1.2) and enable the new feature called "Aditional commands", and add your command in the section "Additional Commands (Before)" (See the image as reference)
![Screenshot 2024-08-14 at 12 32 47 PM](https://github.com/user-attachments/assets/07e82d19-85ec-4766-b728-cb683ccb9826)

**Comment by lajg-dev** [comment]
Please let me know if now it is working for you, so I could close this ticket.

**Comment by jborden** [comment]
Apologies for the late reply, I was out of town when and just missed the update from Github. 

tl;dr - You need to encrypt text at the command line using per item 2 in order to enter your passphrase to GPG. It will cache it. After doing this, everything will work in Obsidian.

Here is how I debugged this:

1. I tried using the `Additional Commands (Before)` with
```
GPG_TTY=$(tty) ; export GPG_TTY
```

This got rid of the 'Inappropriate ioctl for device' message, but I then had a new one 'gpg: signing failed: No such file or directory' 
 

2.  I was trying to use the command line to debug this, so the console logging of the exact command is helpful. So I did a 
```bash
"/opt/homebrew/bin/gpg" --batch --encrypt --armor --trust-model always --sign --local-user 64DD5A914D4419FF --recipient 64DD5A914D4419FF
```
I entered some text, hit Ctrl-D and was prompted for my key pass phrase. I provided it. It encrypted the text. I assume it cached the password because on subsequent terminal calls, it just encrypted the text without further prompting of the password.

3. I was then able to encrypt text in Obsidian, yay! Oddly, I also no longer need the `Additional Commands (Before)` entry anymore. Everything just works. I assume the caching of the passphrase will have some time component to it. I waited until the passphrase was no longer cached and tried to encrypt in Obsidian again. I got the `Inappropriate ioctl for device` again! I entered the passphrase and everything just worked again. In summary, the SO suggestion was a red herring. 

Thanks for the additional work towards fixing this!

**Comment by lajg-dev** [comment]
Thanks for the feedback, I'll close this ticket, if you need further assistance in something related this Obsidian plugin, please let me know

**Comment by Kek5chen** [comment]
It looks like gpg-agent is missing a pinentry program. As I mentioned.

This issue can be solved using doing the following:
```
brew install pinentry-mac
echo "pinentry-program /opt/homebrew/bin/pinentry-mac" > ~/.gnupg/gpg-agent.conf
killall gpg-agent
```
The location of homebrew binaries is different for Intel based macs. Beware.
